In [1]:
#imports

import pandas as pd
import numpy as np

### Question 1
#### What's the version of Pandas that you installed?


In [2]:
pd.__version__

'1.5.2'

### Question 2


In [3]:
#reading the data

df = pd.read_csv("data/housing.csv")

In [4]:
df.sample()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
8054    -118.18     33.85                30.0       2548.0           717.0   

      population  households  median_income  median_house_value  \
8054      2086.0       700.0         0.7007            134400.0   

     ocean_proximity  
8054       <1H OCEAN

#### How many columns are in the dataset?

In [5]:
df.shape

(20640, 10)

### Answer: 10

### Question 3

### Which columns in the dataset have missing values?

In [6]:
df.isna().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

### Answer: total_bedrooms

### Question 4

#### How many unique values does the ocean_proximity column have?

In [7]:
df.ocean_proximity.nunique()

5

### Answer:  5


### What's the average value of the median_house_value for the houses located near the bay?

In [8]:
df.median_house_value.mean()

206855.81690891474

### Answer:  average value is 206855.81690891474

### Question 6
### Calculate the average of total_bedrooms column in the dataset.


In [10]:
df.total_bedrooms.mean()

537.8705525375618

### Use the fillna method to fill the missing values in total_bedrooms with the mean value from the previous step.


In [11]:
df.total_bedrooms.isna().sum()

207

In [12]:
df['total_bedrooms'] = df.total_bedrooms.fillna(537.8705525375618)

Now, calculate the average of total_bedrooms again.
Has it changed?

Hint: take into account only 3 digits after the decimal point.

In [13]:
df.total_bedrooms.mean()

537.8705525375617

In [34]:
#df.total_bedrooms.isna().sum()

0

### Question 7
#### Select all the options located on islands.
#### Select only columns housing_median_age, total_rooms, total_bedrooms.


In [14]:
df.columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value', 'ocean_proximity'],
      dtype='object')

In [15]:
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [16]:
df.ocean_proximity.unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

In [17]:
df.query("ocean_proximity == 'ISLAND'").shape

(5, 10)

In [18]:
df2 = df.query("ocean_proximity == 'ISLAND'")

In [19]:
df2 = df2[["housing_median_age", 'total_rooms', 'total_bedrooms']]

In [20]:
#Get the underlying NumPy array. Let's call it X.
X = df2.to_numpy()

In [21]:
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [22]:
tx = X.transpose() #or X.T for same results

In [23]:
tx

array([[  27.,   52.,   52.,   52.,   29.],
       [1675., 2359., 2127.,  996.,  716.],
       [ 521.,  591.,  512.,  264.,  214.]])

In [24]:
#Compute matrix-matrix multiplication between the transpose of X and X
#same as below

#or np.dot(tx, X)
np.matmul(tx,X)


array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [25]:
#Compute matrix-matrix multiplication between the transpose of X and X
xtx = tx@X

In [26]:
xtx

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [27]:
#Compute the inverse of XTX.
inv_xtx = np.linalg.inv(xtx)

In [28]:
inv_xtx

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [29]:
#Create an array y with values

y = [950, 1300, 800, 1000, 1300]

In [30]:
#7 Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

result = inv_xtx@tx

In [31]:
w = result@y

In [33]:
w

array([23.12330961, -1.48124183,  5.69922946])

In [35]:
#8 What's the value of the last element of w?
w[-1]

5.699229455065586